File: ordered_list/ordered_list.py
class Node:
    def __init__(self, v):
        self.value = v
        self.prev = None
        self.next = None


class OrderedList:
    def __init__(self, asc):
        self.head = None
        self.tail = None
        self.__ascending = asc

    def compare(self, v1, v2):
        if v1 < v2:
            return -1 if self.__ascending else 1
        if v1 == v2:
            return 0
        return 1 if self.__ascending else -1

    def add(self, value):
        new_node = Node(value)

        if not self.head:
            self.head = new_node
            self.tail = new_node
            return
        node = self.head
        while node:
 
            if not node.next and self.compare(node.value, new_node.value) < 0:
                
                node.next = new_node
                new_node.prev = node
                self.tail = new_node
                return
            if not node.next and self.compare(node.value, new_node.value) >= 0:
                node.prev = new_node
                self.tail = node
                new_node.next = node
                if self.head == node:
                    self.head = new_node
                return

            if (
                node.next
                and self.compare(node.value, new_node.value) < 0
                and self.compare(node.next.value, new_node.value) >= 0
            ):
                node.next.prev = new_node
                new_node.next = node.next
                node.next = new_node
                new_node.prev = node
                return

            if self.compare(node.value, new_node.value) >= 0:
                node.prev = new_node
                new_node.next = node
                self.head = new_node
                return

            node = node.next

    def find(self, val):
        node = self.head
        while node:
            if node.value == val:
                return node
            if self.__ascending and node.value > val:
                return None
            if not self.__ascending and node.value < val:
                return None
            node = node.next
        return None

    def delete(self, val):
        if not self.head:
            return None

        if self.head.value == val:
            self.head = self.head.next
            if not self.head:
                self.tail = None
                return
            self.head.prev = None
            return

        node = self.head
        if self.head.prev:
            self.head.prev = None

        while node:
            if node.value == val and not node.next:
                    self.tail = node.prev
                    node.prev.next = None
                    return
            if node.value == val:
                node.prev.next = node.next
                node.next.prev = node.prev
                return
            node = node.next
        return None

    def clean(self, asc):
        self.__ascending = asc
        self.head = None
        self.tail = None

    def len(self):
        res = 0
        node = self.head
        while node:
            res += 1
            node = node.next
        return res

    def get_all(self):
        r = []
        node = self.head
        while node != None:
            r.append(node)
            node = node.next
        return r


class OrderedStringList(OrderedList):
    def __init__(self, asc):
        super(OrderedStringList, self).__init__(asc)
        self.__ascending = asc

    def compare(self, v1, v2):
        v1, v2 = str(v1).strip(), str(v2).strip()
        if v1 < v2:
            return -1 if self.__ascending else 1
        if v1 == v2:
            return 0
        return 1 if self.__ascending else -1

File: asd_go/orderedlist/ordered_list.go
package main

import (
	"constraints"
	"errors"
	"os"
)

type Node[T constraints.Ordered] struct {
	prev  *Node[T]
	next  *Node[T]
	value T
}

type OrderedList[T constraints.Ordered] struct {
	head       *Node[T]
	tail       *Node[T]
	_ascending bool
}

func (l *OrderedList[T]) Count() int {
	count := 0
	node := l.head
	for node != nil {
		count++
		node = node.next
	}
	return count
}

func (l *OrderedList[T]) Add(item T) {
	newNode := &Node[T]{value: item}
	if l.head == nil {
		l.head = newNode
		l.tail = newNode
		return
	}
	n := l.head
	for n != nil {
		if (l._ascending && l.Compare(n.value, item) >= 0) || (!l._ascending && l.Compare(n.value, item) <= 0) {
			l.addToMiddle(n, newNode)
			return
		}
		n = n.next
	}
	newNode.prev = l.tail
	l.tail.next = newNode
	l.tail = newNode
}

func (l *OrderedList[T]) addToMiddle(n, newNode *Node[T]) {
	if n == l.head {
		newNode.next = l.head
		l.head.prev = newNode
		l.head = newNode
	} else {
		newNode.prev = n.prev
		n.prev.next = newNode
		n.prev = newNode
		newNode.next = n
	}
}

func (l *OrderedList[T]) Find(n T) (Node[T], error) {
	node := l.head
	for node != nil {
		if (l._ascending && l.Compare(node.value, n) > 0) || (!l._ascending && l.Compare(node.value, n) < 0) {
			return Node[T]{}, errors.New("node not found")
		}
		if l.Compare(node.value, n) == 0 {
			return *node, nil
		}
		node = node.next
	}
	return Node[T]{}, errors.New("node not found")
}

func (l *OrderedList[T]) Delete(n T) {
	node := l.head

	for node != nil {
		if l.Compare(node.value, n) != 0 {
			node = node.next
			continue
		}

		if node == l.head {
			l.head = l.head.next
		}

		if node.prev != nil {
			node.prev.next = node.next
		}

		if node.next != nil {
			node.next.prev = node.prev
		} else {
			l.tail = node.prev
		}

		if l.head == nil {
			l.tail = nil
		}

		return
	}
}

func (l *OrderedList[T]) Clear(asc bool) {
	l._ascending = asc
	l.head, l.tail = nil, nil
}

func (l *OrderedList[T]) Compare(v1 T, v2 T) int {
	if v1 < v2 {
		return -1
	}
	if v1 > v2 {
		return +1
	}
	return 0
}